lib/vclib/__init__.py

```diff
@@ -427,26 +432,41 @@ def __init__(self, temp1, temp2, info1=None, info2=None, diff_cmd="diff", diff_o
         if info1 and info2:
             args.extend(["-L", self._label(info1), "-L", self._label(info2)])
         args.extend([temp1, temp2])
+        # We assume pipe buffer for stderr is enough for diff utility,
+        # otherwise, it may cause deadlock.
         self.proc = subprocess.Popen(
-            args, stdout=subprocess.PIPE, bufsize=-1, close_fds=(sys.platform != "win32")
+            args, stdout=subprocess.PIPE, stderr=subprocess.PIPE,
+            encoding="utf-8", errors="surrogateescape",
+            bufsize=-1, close_fds=(sys.platform != "win32")
         )
-        if not isinstance(self.proc.stdout, io.TextIOBase) and isinstance(
-            self.proc.stdout, io.BufferedIOBase
-        ):
-            self.fp = io.TextIOWrapper(self.proc.stdout, encoding="utf-8", errors="surrogateescape")
-        else:
-            self.fp = self.proc.stdout
 
-    def read(self, bytes):
-        return self.fp.read(bytes)
+    def read(self, buf_size):
+        buf = self.proc.stdout.read(buf_size)
+        if buf == "":
+            errs = self.proc.stderr.read()
+            ret = self.proc.poll()
+            if ret not in (None, 0, 1) or errs:
+                if ret is None:
+                    ret = -1
+                raise ExternalDiffError(ret, errs)
+        return buf
 
     def readline(self):
-        return self.fp.readline()
+        buf = self.proc.stdout.readline()
+        if buf == "":
+            errs = self.proc.stderr.read()
+            ret = self.proc.poll()
+            if ret not in (None, 0, 1) or errs:
+                if ret is None:
+                    ret = -1
+                raise ExternalDiffError(ret, errs)
```

Comment:
The identical blocks of code here are begging to be abstracted into a helper function (`_check_process_errors`, or something) with some helpful comments.

If I understand them correctly, after each read from the diff process, we expect to find a return value of 0, 1 (both of which occur when the process has completed, or None (when the process is still running) and to find nothing delivered to stderr from the diff process.  If we get a wayward return value or any errors, we raise.  Right?  I'm curious why a return code of 1 isn't considered errorful, though.  Or is that we just don't care in our situation because we're primarily focused on the output from the diff binary?

Reply:
Thank you for the review.

> The identical blocks of code here are begging to be abstracted into a helper function (_check_process_errors, or something) with some helpful comments.

I'll do.

> If I understand them correctly, after each read from the diff process, we expect to find a return value of 0, 1 (both of which occur when the process has completed, or None (when the process is still running) and to find nothing delivered to stderr from the diff process. If we get a wayward return value or any errors, we raise. Right?

Yes, that's right.

> I'm curious why a return code of 1 isn't considered errorful, though. Or is that we just don't care in our situation because we're primarily focused on the output from the diff binary?

[The exit code of diff utility is specified in POSIX](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/diff.html):

> EXIT STATUS
> 
>     The following exit values shall be returned:
> 
>      0
>         No differences were found.
>      1
>         Differences were found.
>     >1
>         An error occurred.

It is also said "The standard error shall be used only for diagnostic messages."  So yes, it seems we could care output of stderr only. Or we could wait for exiting diff process, with some short timeout (As far as I tested with diff -U-1 on GNU diff 3.8, `ret` is always 2).

Reply:
Thank you. I'll merge it